rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+# Reducing Primitive Creation Overhead for Dynamic Shapes
+
+## Motivation
+
+oneDNN API requires users to specify tensors dimensions to create a primitive,
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
```

Comment:
Is there kernel performance impact?

Reply:
There shouldn't be any kernel performance impacts. At a high level, oneDNN does the following transformation sequence:
`problem data ->  optimization set  -> JIT code generator -> optimized kernel`. Under this guise, the existing primitive cache is a tool to short circuit this process allow a direct jump from `problem data -> optimized function`. The kernel cache, similarly, is a tool to jump from `optimization set -> optimized kernel` to avoid the `JIT code generator` cost. Internally, we can then make sure the possible optimization sets are, in some sense, minimal so that cache hits are likely.

Reply:
I see. I thought, before that, oneDNN does sth like,
`problem data1 -> primitive1 -> opt_kernel1`
`problem data2 -> primitive2 -> opt_kernel2`
But as your comments, oneDNN does sth like,
`problem data1 -> primitive1 -> opt_kernel1`
`problem data2 -> primitive2 -> opt_kernel1`
So it makes sense after the proposal,
`problem data1 -> opt_kernel1`
`problem data2 -> opt_kernel1`
And kernel performance should be no difference.